src/NUnitFramework/framework/Attributes/RetryAttribute.cs

```diff
@@ -91,10 +91,10 @@ public override TestResult Execute(TestExecutionContext context)
 
                 while (count-- > 0)
                 {
-                    context.CurrentResult = innerCommand.Execute(context);
+                    innerCommand.ExecuteAndSetCurrentResult(context);
 
-                    if (context.CurrentResult.ResultState != ResultState.Failure)
-                        break;
+                    if (context.CurrentResult.ResultState.Status != TestStatus.Failed) break;
+                    if (context.CurrentResult.ResultState.Label == ResultState.Error.Label) break;
```

Comment:
This is odd when you read it. I guess you're avoiding use of the string but we use them elsewhere. Should we have manifest constants for the known strings? 

Still I'm OK with this if you plan to come back to it right away on the general issue. But if itd likely to sit for a while can we make this a but clearer?

Reply:
Maybe. Would an `ResultState.ErrorLabel` or `ResultStateLabels.Error` constant be any clearer or nicer?
I can do `"Error"` too if that's you preference and it's more consistent with the codebase that way.

Reply:
This is a semantic change, right? Retry will now Retry on both failure and error? Or has this functionality just moved?

I have no problem with that if it is, just want to understand what the change is. 🙂

Reply:
No, this code avoids the semantic change. The existing tests enforcing no retry on error are untouched and the tests I added follow the same semantics.
This line makes sure retry still does not happen if there is an error:
```cs
if (context.CurrentResult.ResultState.Label == ResultState.Error.Label) break;
```

Reply:
The reason why it's confusing to look at the diff is that the original line does not do what it looks like it does.

Reply:
I think I'm a bit confused as to where the original bug was! 😆 Can you clarify why both checks are needed?

Reply:
Oh, does `Error` come under `TestStatus.Failed`?

Reply:
Each of the three "fixes" commits fixes a separate bug. :D Looking commit-by-commit will probably be enlightening because I tried to keep each one simple.
Yes, `Error` is `TestStatus.Failed` and `Label = "Error"`.

Reply:
Well I'm glad you've opened an issue about making this simpler! 😄

Reply:
@Chris Now that you mention it, it would probably be a good idea to make `Execute` protected now rather than later and make everyone use it. Right now there are three kinds of failure every delegating test command must handle:

- innerCommand.Execute returning Failed without "Error" label
- innerCommand.Execute returning Failed with "Error" label
- innerCommand.Execute throwing or letting an exception bubble through

Having two ways to signify the same thing causes a chill to go down my spine :D that's where bugs spontaneously generate.

Reply:
The three things you list are not three different ways to signify the same thing. They are three different things. Not only that, but you have underestimated how many different things there are... it's not just with and without "Error" label, it's an unlimited number of possible `Label` values. We currently use empty (a simple failure), "Error", "Cancelled" and "Invalid". We have discussed adding a few other internal values and users are expected to define their own (future) labels.

Reply:
No, but two of the three are. You're right, there are many more. The current system makes it difficult to property implement test commands which need to react differently to different kinds of failure, as RetryAttribute is currently specced to do.

Reply:
There is only one __Failure__ value for ResultState and that's the one that `RetryAttribute` is supposed to repeat on as currently designed. Making a negative test is what opens the door to other values. As written now, it will repeat even if the user cancels the test!

Reply:
Oh! I'm finally seeing this comment. That put it into focus for me. Thanks!